### PR TITLE
objectfile: Remove panics

### PR DESCRIPTION
**Pull request by kakkoyun**
### Why?

Partially addresses https://github.com/parca-dev/parca-agent/issues/1798
Related: https://github.com/parca-dev/parca-agent/issues/1962 and https://github.com/parca-dev/parca-agent/pull/2031

See also https://github.com/parca-dev/parca-agent/pull/2032#pullrequestreview-1635163776

### What?

- [Convert noreleases todos](https://github.com/parca-dev/parca-agent/commit/29c1a951bad9e9bf169257b8505afad4a0afa219)
- Convert panics to errors
- [Use atomic boolean to track life-cycle of the file](https://github.com/parca-dev/parca-agent/commit/9a63101aaf24639df7c1c81097ed394cf61dc627)

### <samp>🤖 Generated by Copilot at 9a63101</samp>

This pull request updates the `libbpf` dependency and fixes some compilation issues. It also refactors the `ObjectFile` and `readMaps` types and functions to improve concurrency, memory management, and testability. It simplifies the `obj.ELF()` function to return only the `*elf.File` and an error, and updates the comments in the `Map` struct to match the code.

### <samp>🤖 Generated by Copilot at 9a63101</samp>

*  Simplify the `obj.ELF()` function to return only the `*elf.File` and an `error`, and use the garbage collector to close the file descriptor instead of a manual `release()` function ([link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-a01e25a5f5c397126a4a459eee2a2c30fb9acbd1a6ea9fc5e32dd3136b978d56L142-R145), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-e34917e1d967c4b448d1cd842e1aace9c273ab0bf4e4ca198186f70d82dbfd74L339-R340), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L275-R278), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L335-R334), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L341), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L444-R442), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L452-R451), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-5c4a0ca9a2747c99b32f629099e552b2582da981ff90f6bcedd6044dbe11e359L476-R476), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-52fbddf5028b2561d9f9b87e8db382657a58ee3e4d7974523e3b07706961bc5fL67-R70), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL222-R225), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-531c87f587fd53526fd4780bd78b14affdc797b4687500fc66cfea88f929957dL402-R405), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-b3a7f294a0fa5a71ab06af71de938e5dfbc119109c6d13ce194e58efec847517L82-R85))
*  Refactor the `ObjectFile` type in `pkg/objectfile/object_file.go` to use an `*atomic.Bool` for the `closed` field, instead of a `bool` protected by a `*sync.RWMutex`, and to use a `*io.SectionReader` for the `file` field, instead of a `*os.File`, to allow concurrent reads ([link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537L25-R27), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537L41-R48), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537L61-R88), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537L110-R112), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-678278e8b8d78d40114e13df0e0e0033d0b9d5375d2e01e85ac0d05598fd9537L125-R119), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-092464fb7162e237d2388d6d49db571affbfb356a538c3490240cfc9ec581b95L24), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-092464fb7162e237d2388d6d49db571affbfb356a538c3490240cfc9ec581b95R30), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-092464fb7162e237d2388d6d49db571affbfb356a538c3490240cfc9ec581b95L291-R292))
*  Replace the comment `@norelease: DRY` with `TODO(kakkoyun): DRY` in the `Makefile` and the `bpf/Makefile` to indicate the code duplication should be refactored by the author of the change ([link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L63-R63), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L165-R165), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-93767fe53d11d3998146fe42c981bd75c59caf52abce71885b6ae808525709faL17-R17), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-93767fe53d11d3998146fe42c981bd75c59caf52abce71885b6ae808525709faL27-R27))
*  Replace the comment `// TODO: DRY` with `// TODO(kakkoyun): DRY` in several files to indicate the code duplication should be refactored by the author of the change ([link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-ef52c94872492944c90e9e3219ef0595be58f11985effb033905912b77f544a0L112-R112), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-1c1bf15e305b51045a2b4c72fc4c1a64fb73ed42c915ab8f0241143e4a76c307L457-R457), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-1c1bf15e305b51045a2b4c72fc4c1a64fb73ed42c915ab8f0241143e4a76c307L484-R484), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-1c1bf15e305b51045a2b4c72fc4c1a64fb73ed42c915ab8f0241143e4a76c307L511-R511), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-1c1bf15e305b51045a2b4c72fc4c1a64fb73ed42c915ab8f0241143e4a76c307L537-R537), [link](https://github.com/parca-dev/parca-agent/pull/2032/files?diff=unified&w=0#diff-eddc84f9aaf62826e76da4882f2f328b5607b6130bf887a194ed0743ce262372L28-R28))

### Test Plan

* Run agent with the race detector over a period of time.

### CODE REVIEW

**Comment by javierhonduco** [comment]
Let's discuss tomorrow but I think that this panic was great because it is a symptom of a bigger issue, the race conditions in some core parts of the Agent. I would suggest first fixing them and working on understanding all the failures in-depth before making changes here that might cover up issues.

**Comment by kakkoyun** [comment]
> Let's discuss tomorrow but I think that this panic was great because it is a symptom of a bigger issue, the race conditions in some core parts of the Agent. I would suggest first fixing them and working on understanding all the failures in-depth before making changes here that might cover up issues.

Happy to discuss but I spent some time on this yesterday and this panic is unavoidable without implementing proper reference counting. And the failures are negligible.

**Comment by javierhonduco** [comment]
> Happy to discuss but I spent some time on this yesterday and this panic is unavoidable without implementing proper reference counting. 

Not sure I understand why reference counting is needed. My understanding is that if we rely on the map as a backing store and remove old entries, everything should work as we expect. Happy to chat because I am sure I am missing something!

> And the failures are negligible.

I would say we don't really now at the moment. We just added instrumentation for this and it will take a while for most of our users to update to this revision (+is not implemented in ps, yet)

**Comment by javierhonduco** [comment]
Another thing to bear in mind here is that the object pool at the minute is rather large, which helps reduce the number of crashes we see

File: pkg/objectfile/object_file.go [hunk 1]
```diff
@@ -107,23 +98,25 @@ func (o *ObjectFile) close() error {
...
+	if o.closed.Load() {
 		return errors.Join(ErrAlreadyClosed, fmt.Errorf("file %s is already closed by: %s", o.Path, frames(o.closedBy)))
 	}
-
+	o.closed.Store(true)
+	// NOTICE: This close is a no-op. The elf.File is opened through elf.NewFile,
+	// which does not initialize a closer. It's here because of testing purposes.
+	// Because of this the underlying file descriptor will be closed by the GC.
+	// If there is an active reader, it will conclude successfully.
+	// Only downside will be to re-opening the file if the ObjectFile is evicted
```
Comment (javierhonduco) [inline pkg/objectfile/object_file.go:111]: As mentioned in the chat today, this will most likely change in the future

**Review by javierhonduco — APPROVED** [review]
After having a chat with Kemal, I am convinced this change makes a lot of sense right now. Just to summarise some key points of our conversation:
- Ideally, in the future, we want to evolve this part of the code to not have deal with closed files when we keep a reference to them. Right now we mark the FD as closed in when the LRU cache evicts the object;
- This PR is a stepping stone towards the final solution. Rather than panic, returning an error makes sense as we just mark the file as closed anyway. 

Personally, I really like the simplification of the code here, getting rid of `release()` is a good idea

**Pull request merged**